File: matlab_reference/N-ret-mod---Source-main/certainty_function_definer.m
function cert_fun = certainty_function_definer(cert_fun_choice)

    % FRAFA APRIL 2023
    if strcmp(cert_fun_choice,'FRAFA_apr2023')
        cert_fun = @(dist,range,width,sill) (dist<=width).*sill+(dist>width).*sill.*(exp(-(3*power((dist-width),2))./(power(range,2))));
 
    % ILM SEP 2023
    elseif strcmp(cert_fun_choice,'ILM_sep2023')
        cert_fun = @(dist,range,width,sill) (dist<=width).*sill.*(exp(-(3*power((dist-width),2))./(power(range,2))));

    % ILM OCT 2023
    elseif strcmp(cert_fun_choice,'ILM_oct2023')
        cert_fun = @(dist,range,width,sill) (dist<=width).*sill.*(exp(-(3*power((dist),2))./(power(range,2))));

    % RBM OCT 2023
    elseif strcmp(cert_fun_choice,'RBM_oct2023')
        cert_fun_rbm_oct_ins = @(dist,range,width,sill,damp) sill.*(exp(-(damp*power((dist),2))./(power(range,2))));
        cert_fun_rbm_oct_out = @(dist,range,width,sill) sill.*(exp(-(3*power((dist),2))./(power(range,2))));

        cert_fun = @(dist,range,width,sill) (dist<=width).*cert_fun_rbm_oct_ins(dist,range,width,sill,1.5)+...
                                                 (dist>width).*cert_fun_rbm_oct_out(dist,range,width,sill+(sill-cert_fun_rbm_oct_ins(width,range,width,sill,1.5)));
    end

File: certainty_function_definer.py
import numpy as np

def certainty_function_definer(cert_fun_choice):
    if cert_fun_choice == 'FRAFA_apr2023':
        def cert_fun(dist, range_, width, sill):
            return np.where(dist <= width, sill, sill * np.exp(-3 * (dist - width)**2 / range_**2))
    
    elif cert_fun_choice == 'ILM_sep2023':
        # Multiplicative MATLAB form (dist<=width).*sill.*exp(...), NOT
        # np.where(..., 0): at a no-data cell dist (or range) is NaN, and
        # 0*NaN=NaN propagates to the NODATA sentinel just like MATLAB's
        # Grid(isnan)=100000 — whereas np.where would leak a literal 0 -> 1/0=+inf.
        def cert_fun(dist, range_, width, sill):
            return (dist <= width) * sill * np.exp(-3 * (dist - width)**2 / range_**2)

    elif cert_fun_choice == 'ILM_oct2023':
        def cert_fun(dist, range_, width, sill):
            return (dist <= width) * sill * np.exp(-3 * dist**2 / range_**2)
    
    elif cert_fun_choice == 'RBM_oct2023':
        def cert_fun_rbm_oct_ins(dist, range_, width, sill, damp):
            return sill * np.exp(-damp * dist**2 / range_**2)

        def cert_fun_rbm_oct_out(dist, range_, width, sill):
            return sill * np.exp(-3 * dist**2 / range_**2)

        def cert_fun(dist, range_, width, sill):
            inside = dist <= width
            outside = dist > width
            sill_adjusted = sill + (sill - cert_fun_rbm_oct_ins(width, range_, width, sill, 1.5))
            return np.where(inside,
                            cert_fun_rbm_oct_ins(dist, range_, width, sill, 1.5),
                            cert_fun_rbm_oct_out(dist, range_, width, sill_adjusted))
    else:
        raise ValueError(f"Unknown certainty function choice: {cert_fun_choice}")

    return cert_fun